Added VSCode debugging (for MvcSandbox) & code analysis support

ShreyasJejurkar opened the pull request:
This PR is a part of ongoing efforts for improving the experience for VSCode users when working with the aspnetcore repo. 

### PR includes 
1. Debugging support for the <b>MvcSandbox</b> project. 
2. Roslyn code analysis support for the entire repo including Public API custom analyzers & code fixes. 

### Changes 
1. Added `startvscode.cmd` at the root folder and for the Mvc folder.
2. Added `omnisharp.json` for root and for the Mvc folder with the required configuration. 


Contributes to https://github.com/dotnet/aspnetcore/issues/27260 https://github.com/dotnet/aspnetcore/issues/27380 https://github.com/dotnet/aspnetcore/issues/27609

Please check out the following video snippets about the experience!

## Debugging MvcSandbox project 

https://user-images.githubusercontent.com/17148381/105345913-dd782900-5c0a-11eb-92fe-62f2f036b66c.mp4


## Roslyn analyzer 
https://user-images.githubusercontent.com/17148381/105340687-3a241580-5c04-11eb-95fb-570b13361dfe.mp4

ShreyasJejurkar committed:
Added VSCode debugging support for MvcSandbox

ShreyasJejurkar committed:
Added ignored vscode folders

pranavkm commented on .vscode/launch.json:
```diff
@@ -1,17 +1,22 @@
+    "type": "coreclr",
+    "preLaunchTask": "MvcSandbox-build",
+    "request": "launch",
+    "program": "${workspaceFolder}/artifacts/bin/MvcSandbox/Debug/net6.0/MvcSandbox.dll",
```
Having the target framework here is problematic. We'd have to remember to update this every time we rev the TFM and it's easy to miss. Is there an alternative to this?

pranavkm replied:
Also, I wouldn't expect us to have any MVC specific tasks or actions outside of the MVC folder. We rarely expect developers to open up the solution in entirety.

captainsafia replied:
Unfortunately, the C# extension for VS Code requires that you provide a full-path to the assembly in order to debug it.

VS Code only picks up the launch config in the root `.vscode` directory of the workspace. So we can move this config to the analyzers project, but it would mean that users would have to open that specific folder to load the config. That's probably a safe assumption to make.

MCCshreyas replied:
@pranavkm I guess we can do that is there any env variable or path variable that we are using for `TargetFramework` in the aspnetcore repo, so I will check if VScode can pick up that so that we don't have to hardcode TFW everytime!

MCCshreyas replied:
@captainsafia I am finding a way (if there any) to utilize a single `.vscode` folder which is at the root. So although if anyone opens a subfolder of the repo, they should see debug options from that `vscode` folder. 
Right now, in this PR I have added `.vscode` folder in root as well and in the most commonly use `Mvc` folder.

MCCshreyas replied:
I am finding a way to read this value in tasks.json so we can utilize that there. 

https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/eng/Versions.props#L45

captainsafia replied:
> I am finding a way to read this value in tasks.json so we can utilize that there.

If it's feasible to read from that property file (or maybe an environment variable), in the `launch.json` file that would be great!

> I am finding a way (if there any) to utilize a single .vscode folder which is at the root. So although if anyone opens a subfolder of the repo, they should see debug options from that vscode folder.

Got it. Hopefully, there's a strategy for this. I'm not aware of anything off the top of my head.

Thanks for taking the time to work on this!

MCCshreyas replied:
> Got it. Hopefully, there's a strategy for this. I'm not aware of anything off the top of my head.

Initial research on this found that it's not possible even today! Seems like we need to have a separate `.vscode` folder for samples OR for most likely projects which we all run on a regular basis for development for example `MvcSandbox`. 

> Thanks for taking the time to work on this!

Oh, please! My intension is to help all of you guys with simple bugs or problems as much as I can so that you can focus on bigger features! Hehe 😁😁

captainsafia replied:
@MCCshreyas I figured as much. For now, it's fine to place a .vscode directory where ever there is a `build.cmd` file.

Thanks for your help! Much appreciated!

MCCshreyas replied:
I will get that added and tested in this PR in upcoming commits!

Thanks! 🙌

captainsafia replied:
Considering that we can now read the target framework from an environment variable, let's update this to `${env:TARGET}`.

pranavkm commented on src/Mvc/omnisharp.json:
```diff
@@ -0,0 +1,18 @@
+    "RenameInStrings": true
+  },
+  "msbuild": {
+    "MSBuildSDKsPath": "..\\..\\.dotnet",
```
We usually expect to use the `dotnet` in the path. Can we not specify this?

MCCshreyas replied:
This points to preview SDK that has been downloaded in the root `.dotnet` folder!

pranavkm replied:
Sure, but with VS you assume you've launched it with startvs or run activate so that the dotnet that's in the path is the right one. We should do the same here.

MCCshreyas replied:
yeah, just like we have instructed people to use `startvs.cmd` in the same way we will ask people to open repo in VSCode using `startvscode.cmd` which I have included in the PR as well!

ShreyasJejurkar committed:
1. Convered some more projects.
2. Added env variable for TargetFramework and used it in launch.json file rather than
hardcoding TFW.

MCCshreyas commented on startvscode.cmd:
```diff
@@ -0,0 +1,32 @@
+SET PATH=%DOTNET_ROOT%;%PATH%
+
+:: Sets TFW for Visual Studio Code usage
+SET TARGET=net6.0
```
This is the only place where we need to change `TargetFramework` when it's gets changed next time. All child `startvscode.cmd` file will read this value!

captainsafia replied:
Great! Were you able to figure out a way to incorporate this into the target framework referenced in the `launch.json` files?

MCCshreyas replied:
Yes Safia, `launch.json` will use this `env` value. As you can see below. 
https://github.com/dotnet/aspnetcore/blob/c7a5c9e68f2a43b669242aad4028729cdce1ab98/src/Mvc/.vscode/launch.json#L9

So, at the time of the next release, we just need to change the `TFW` in `startvscode.cmd` file.

captainsafia commented on startvscode.cmd:
```diff
@@ -0,0 +1,32 @@
+@ECHO OFF
```
Can you implement a Shell-script based variant of this a la `startvscode.sh` for users on mac/Linux?

Let me know if you need help with this.

MCCshreyas replied:
Ummm, need to check. I will try to do it this weekend!

MCCshreyas replied:
Hi @captainsafia, Sorry, am not much familiar with MacOS and it's shell working! 

If you managed to get time to do that and wanted to get that done in this PR only, I think you can push commit to this PR as well! Otherwise better to have a follow up PR after this! 😇

captainsafia replied:
Got it -- we can address this in a follow-up PR.

MCCshreyas replied:
Very thanks @captainsafia! 😇

captainsafia commented on startvscode.cmd:
```diff
@@ -0,0 +1,32 @@
+@ECHO OFF
+SETLOCAL
```
Let's update the `BuildFromSource.md` file to indicate that users will need to run this script before launching their projects in VS Code.

MCCshreyas replied:
I have updated `BuildFromSource.md`, but please let me know if there are more details to put on!

captainsafia commented on src/Mvc/.vscode/extensions.json:
```diff
@@ -0,0 +1,7 @@
+{
```
It's strange that we only have this here and not in the other directories.

Would recommend removing it and including notes in the `BuildFromSource.md` about what extensions we recommend people use.

MCCshreyas replied:
Deleted `extensions.json` and added instructions to `BuildFromSource.md` for recommended extension!

captainsafia reviewed (commented):
Left some comments inline. I think we might want to go ahead and make a list of the projects that need VS Code support added so that we verify what we've covered.

ShreyasJejurkar committed:
Addressed PR feedback.
1. Removed extensions.json
2. Updated BuildFromSource.md for Vscode instructions.
3. Added launch settings for BasicTestApp.

MCCshreyas commented on src/Components/.vscode/launch.json:
```diff
@@ -0,0 +1,39 @@
+      "env": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      },
+    }
```
@captainsafia Yes. Added in the latest commit! Thanks!

ShreyasJejurkar committed:
Merge branch 'main' into vscode-support

captainsafia commented on .vscode/launch.json:
```diff
@@ -1,17 +1,22 @@
-      "request": "launch",
-      "name": "ps: Interactive Session",
-      "cwd": "${workspaceRoot}"
+  {
```
Since we already have this launch script defined under the Mvc directory we can revert this change.

MCCshreyas replied:
Reverted!

captainsafia commented on .vscode/tasks.json:
```diff
@@ -0,0 +1,15 @@
+  "version": "2.0.0",
+  "tasks": [
+      {
+          "label": "MvcSandbox-build",
```
Ditto here. Let's move this task to the `Mvc` directory.

MCCshreyas replied:
Removed this task entry!

captainsafia commented on docs/BuildFromSource.md:
```diff
@@ -95,6 +95,14 @@ The build should find any JDK 11 or newer installation on the machine as long as
 
+#### Visual Studio Code Extension
+
+Following extension are recommanded for Visual Studio Code in order to work with ASP.NET Core repository.
```
```suggestion
The following extensions are recommended when developing in the ASP.NET Core repository with Visual Studio Code.
```

captainsafia commented on src/Mvc/.vscode/settings.json:
```diff
@@ -0,0 +1,9 @@
+{
```
Let's remove this `settings.json` file for now and retain a `launch.json` and `tasks.json` file in each of the subdirectories.

captainsafia reviewed (commented):
Looking good. Left some final comments inline then we can merge.

ShreyasJejurkar committed:
Addressed PR feedback!

ShreyasJejurkar committed:
Merge branch 'vscode-support' of https://github.com/MCCshreyas/aspnetcore into vscode-support

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore into vscode-support
